core/evtforward/ethcall/result.go

```diff
@@ -49,18 +51,43 @@ func newResult(call Call, bytes []byte) (Result, error) {
 }
 
 func normaliseValues(values []any, rules map[string]string) (map[string]string, error) {
+	// The data in 'values' is relatively well typed, after being unpacked by the ABI.
+	// For example, a uint256 will be a big.Int, structs returned from the contract call
+	// will be anonymous go structs rather than a string map etc.. In order to fish data
+	// out with jsonpath it need to be simple lists, maps, strings and numbers; so we
+	// serialise to json and then deserialise to an into an []interface{}.
+	valuesJson, err := json.Marshal(values)
+	if err != nil {
+		return nil, fmt.Errorf("unable to serialse values: %v", err)
+	}
+
+	valuesSimple := []interface{}{}
+	d := json.NewDecoder(bytes.NewBuffer(valuesJson))
+	// Keep numbers as a json.Number, which holds the original string representation
+	// otherwise all numbers get cast to float64 which is no good for e.g uint256.
+	d.UseNumber()
+	err = d.Decode(&valuesSimple)
+	if err != nil {
+		return nil, fmt.Errorf("unable to deserialse values: %v", err)
+	}
+
 	res := make(map[string]string)
+
 	for key, path := range rules {
-		value, err := jsonpath.Get(path, values)
+		value, err := jsonpath.Get(path, valuesSimple)
 		if err != nil {
 			return nil, fmt.Errorf("unable to normalise key %v: %v", key, err)
 		}
-		valueJson, err := json.Marshal(value)
-		if err != nil {
-			return nil, fmt.Errorf("unable to serialse normalised value: %v", err)
+		switch v := value.(type) {
+		case json.Number:
+			res[key] = v.String()
+		case float64:
+			res[key] = strconv.FormatFloat(v, 'f', -1, 64)
```

Comment:
wouldn't that be eventual source of non determinism, can we instead just get the string of the float64 directly?

Reply:
ah well spotted; so this is for the very specific case where a 'static' value is given in the path itself. any numbers that are in the eth call result get unmarshalled as a json.Number, which is what we want. However, you can give a reques a 'path' to a static value like

```normalisers = {"number_one":1}```

which on the face of it seems rather useless but was actually handy when doing the demo; because as things are set up currently you can only settle a market based on (iirc) a boolean or a timestamp.. 

Our current data processing pipeline has no way for us to actually change the values; only filter them; so if you set up e.g. a termination signal like

```json
                                        "filters": [
                                            {
                                                "key": {
                                                    "name": "myproperty",
                                                    "type": "TYPE_INTEGER"
                                                },
                                                "conditions": [
                                                    {
                                                        "operator": "OPERATOR_GREATER_THAN",
                                                        "value": "100"
                                                    }
                                                ]
                                            },
```

you can't use myproperty as a termination signal, because it is an integer.. so I hacked around that by adding a 'static' json path:

```json
                                        "normalisers": {
                                            "myproperty": "$[0]",
                                            "shouldTerminate": "true"
                                        },
```
and using `shouldTerminate` as the termination property; which works because only if all the filters match does it get broadcast.

That whole thing is a bit grotty, I think we really want some pipeline of 'transformers', so e.g. when we filter we can put the boolean result of the filter as a new property in its own right or something.. but even then, I thought it might be handy to be able to add static values as paths; I dunno maybe an identifier or scaling factor or somthing that is unique to the contract but perhaps not part of the contract call.

so.. back to the actual issue you raised; I don't think it's non deterministic because you only get to the `case float64:` here if the number is generated by the JSONPath itself; all numbers in the ethcall result are kept as strings, wrapped in json.Number because of

```go
	d := json.NewDecoder(bytes.NewBuffer(valuesJson))
	// Keep numbers as a json.Number, which holds the original string representation
	// otherwise all numbers get cast to float64 which is no good for e.g uint256.
	d.UseNumber()
```

Though perhaps I could make that clearer in the code; if you are happy in principle I'll add a comment

Reply:
OK I came up with a different way to do it involving no floats and a bit of poking around in the jsonpath interpreter